### PR TITLE
Add day interval support to sequence function

### PR DESCRIPTION
**Pull request by kewang1024**
Added
- [x]  sequence(date, date, interval day to second)

Followup
 - [ ] [Optimization] Add fast path when step is constant
 - [ ] sequence(date, date, interval year to month)
 - [ ] sequence(timestamp, timestamp, interval day to second)
 - [ ] sequence(timestamp, timestamp, interval year to month)

### CODE REVIEW

**Comment by kewang1024** [comment]
Followup PR: https://github.com/facebookincubator/velox/pull/4739

File: velox/functions/prestosql/Sequence.cpp [hunk 1]
```diff
@@ -131,20 +131,44 @@ class SequenceFunction : public exec::VectorFunction {
...
+      int64_t stop,
+      DecodedVector* stepVector,
+      vector_size_t row,
+      bool isDate) {
+    if (stepVector) {
+      auto step = stepVector->valueAt<int64_t>(row);
+      if (isDate) {
+        VELOX_USER_CHECK(
+            step % kMillisInDay == 0,
+            "sequence step must be a day interval if start and end values are dates");
```
Comment (arunthirupathi) [inline velox/functions/prestosql/Sequence.cpp:164]: Include the step value, row value  along with this error.

File: velox/functions/prestosql/Sequence.cpp [hunk 2]
```diff
@@ -131,20 +131,44 @@ class SequenceFunction : public exec::VectorFunction {
...
       DecodedVector* stopVector,
       DecodedVector* stepVector,
       vector_size_t row) {
     auto start = startVector->valueAt<T>(row);
     auto stop = stopVector->valueAt<T>(row);
-    const int64_t step = (stepVector == nullptr)
-        ? (toInt64(stop) >= toInt64(start) ? 1 : -1)
-        : toInt64(stepVector->valueAt<T>(row));
+    auto step = getStep(
+        toInt64(startVector->valueAt<T>(row)),
```
Comment (arunthirupathi) [inline velox/functions/prestosql/Sequence.cpp:143]: can this use start and stop instead of recomputing this value again.

**Review by arunthirupathi — COMMENTED** [review]
Don't know enough to approve this :(

File: velox/functions/prestosql/Sequence.cpp (hunk 1 above)
Comment (kewang1024) [reply velox/functions/prestosql/Sequence.cpp:164]: Currently I'm keeping the same behavior with Java version, thus the same wording

File: velox/functions/prestosql/Sequence.cpp [hunk 3]
```diff
@@ -131,20 +131,39 @@ class SequenceFunction : public exec::VectorFunction {
...
+      int64_t stop,
+      DecodedVector* stepVector,
+      vector_size_t row,
+      bool isDate) {
+    if (stepVector) {
+      auto step = stepVector->valueAt<int64_t>(row);
+      if (isDate) {
+        VELOX_USER_CHECK(
+            step % kMillisInDay == 0,
+            "sequence step must be a day interval if start and end values are dates");
```
Comment (spershin) [inline velox/functions/prestosql/Sequence.cpp:159]: @kewang1024 
Move this check to checkArguments()?

File: velox/functions/prestosql/Sequence.cpp [hunk 4]
```diff
@@ -77,12 +77,11 @@ class SequenceFunction : public exec::VectorFunction {
     auto rawSizes = sizes->asMutable<vector_size_t>();
     auto rawOffsets = offsets->asMutable<vector_size_t>();
 
+    const bool isDate = args[0]->type() == DATE();
```
Comment (laithsakka) [inline velox/functions/prestosql/Sequence.cpp:80]: does DATE() here construct a shared ptr? I wonder if we shall compare kind instead

File: velox/functions/prestosql/Sequence.cpp [hunk 5]
```diff
@@ -131,20 +131,39 @@ class SequenceFunction : public exec::VectorFunction {
...
       elements[i] = add(start, step * i);
     }
   }
+
+  static int64_t getStep(
+      int64_t start,
+      int64_t stop,
+      DecodedVector* stepVector,
+      vector_size_t row,
+      bool isDate) {
```
Comment (laithsakka) [inline velox/functions/prestosql/Sequence.cpp:153]: nit: use early exist to simplify code
https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code 

```
if (!stepVector) {
     return (stop >= start ? 1 : -1);
}
auto step = stepVector->valueAt<int64_t>(row);
if(!isDate){
       return step;
}
  
// Handle Date
VELOX_USER_CHECK(step % kMillisInDay == 0, 
 "sequence step must be a day interval if start and end values are dates");
return step / kMillisInDay;
```

**Review by laithsakka — COMMENTED** [review]
Looks good the logic probably should be reviewed by spark person

**Pull request closed without merging**